src/PureScript/Backend/Optimizer/Tracer/Printer.purs

```diff
@@ -0,0 +1,516 @@
+            , wrapInParens $ fold
+                [ D.words
+                    [ D.text $ "Step"
+                    , D.text $ show idx
+                    ]
+                , case idx == 0, idx == lastIdx of
+                    true, true -> D.text "; Original/Final"
+                    true, _ -> D.text "; Original"
+                    _, true -> D.text "; Final"
+                    _, _ -> mempty
+                ]
+            ]
+        , D.indent $ printBackendExpr step
+        ]
+    foldStep ident lastIdx idx acc step
+      | idx == 0 = renderSingleStep ident lastIdx idx step
+      | otherwise = D.lines [ acc, D.text $ power "-" 15, renderSingleStep ident lastIdx idx step ]
+
+    foldIdent ident acc steps
+      | acc.idx == 0 =
+          { idx: 1
+          , value:
+              foldlWithIndex (foldStep ident (NonEmptyArray.length steps - 1)) mempty $ NonEmptyArray.toArray steps
+          }
+      | otherwise =
+          { idx: acc.idx + 1
+          , value:
+              D.foldWithSeparator (D.break <> D.break <> (D.text $ power "=" 15) <> D.break <> D.break)
+                [ acc.value
+                , foldlWithIndex (foldStep ident (NonEmptyArray.length steps - 1)) mempty $ NonEmptyArray.toArray steps
+                ]
+          }
+
+  _.value $ foldlWithIndex foldIdent { idx: 0, value: mempty } allSteps
+
+printQualifiedIdent :: Qualified Ident -> Doc Void
+printQualifiedIdent = printQualified \(Ident a) -> D.text a
+
+printQualified :: forall a. (a -> Doc Void) -> Qualified a -> Doc Void
+printQualified printA (Qualified mbMod a) =
+  maybe (printA a) (\(ModuleName mod) -> D.text mod <> D.text "." <> printA a) mbMod
+
+printIdent :: Ident -> Doc Void
+printIdent (Ident i) = D.text i
+
+printLevel :: Level -> Doc Void
+printLevel (Level l) = D.text $ show l
+
+printLocal :: Maybe Ident -> Level -> Doc Void
+printLocal mbIdent lvl = do
+  let lvl' = D.text "v" <> printLevel lvl
+  maybe lvl' (\i -> printIdent i <> D.text "@" <> lvl') mbIdent
+
+printBackendExpr :: BackendExpr -> Doc Void
+printBackendExpr = case _ of
+  ExprSyntax _ b -> printBackendSyntax b
+  ExprRewrite _ b -> printBackendRewrite b
+
+printCurriedApp :: Doc Void -> Doc Void -> Doc Void
+printCurriedApp fn arg =
+  D.text "(" <> D.words [ fn, arg ] <> D.text ")"
+
+printIndentedParens :: Doc Void -> Doc Void
+printIndentedParens = printIndentedParens' mempty
+
+printIndentedParens' :: Doc Void -> Doc Void -> Doc Void
+printIndentedParens' header body =
+  D.lines
+    [ D.text "(" <> header
+    , D.indent body
+    , D.text ")"
+    ]
+
+wrapInParens :: Doc Void -> Doc Void
+wrapInParens = append (D.text "(") <<< flip append (D.text ")")
+
+printLet :: String -> Doc Void -> Doc Void -> Doc Void
+printLet letKeyword identifier binding =
+  D.lines
+    [ D.words
+        [ D.text letKeyword
+        , identifier
+        , D.text "="
+        ]
+    , D.indent $ binding <> D.text ";"
```

Comment:
The goal of printing let this way is to produce an output like below. The usage of `;` was to make it clearer when one ended:
```
let ident1 = 1;
let ident2 = 2;
1 + 2
```
I'm not sure the current code achieves this in all cases (e.g. `RewriteInline` situations), but that was my intent.